File: DataStructures/sorts/merge.js
function merge(list, start, mid, end) {
    const lenArray1 = mid - start + 1;
    const array1 = new Array(lenArray1)
    for (let i = 0; i < lenArray1; i++) {
        array1[i] = list[i + start]
    }
    const lenArray2 = end - mid;
    const array2 = new Array(lenArray2)
    for (let i = 0; i < lenArray2; i++) {
        array2[i] = list[i + mid + 1]
    }

    let leftPointer = 0,
        rightPointer = 0,
        originalPointer = start;
    while (leftPointer < lenArray1 && rightPointer < lenArray2) {
        if (array1[leftPointer] < array2[rightPointer]) {
            list[originalPointer] = array1[leftPointer]
            leftPointer++
        } else {
            list[originalPointer] = array2[rightPointer]
            rightPointer++
        }
        originalPointer++
    }
    while (leftPointer < lenArray1) {
        list[originalPointer] = array1[leftPointer]
        leftPointer++
        originalPointer++
    }
    while (rightPointer < lenArray2) {
        list[originalPointer] = array2[rightPointer]
        rightPointer++
        originalPointer++
    }
}

function mergeSort(arr, start = 0, end = arr.length - 1) {
    if (start < end) {
        const mid = Math.floor((start + end) / 2)
        mergeSort(arr, start, mid)
        mergeSort(arr, mid + 1, end)
        merge(arr, start, mid, end)
    }
    return arr

}

File: DataStructures/sorts/go/merge.go
package main

func merge(list *[]int, start, mid, end int) {
	lengthLeft := mid - start + 1
	leftCopy := make([]int, lengthLeft)
	for i := 0; i < lengthLeft; i++ {
		leftCopy[i] = (*list)[i+start]
	}

	lengthRight := end - mid
	rightCopy := make([]int, lengthRight)
	for i := 0; i < lengthRight; i++ {
		rightCopy[i] = (*list)[i+mid+1]
	}

	left := 0
	right := 0
	original := start
	for left < lengthLeft && right < lengthRight {
		if leftCopy[left] < rightCopy[right] {
			(*list)[original] = leftCopy[left]
			left++
		} else {
			(*list)[original] = rightCopy[right]
			right++
		}
		original++
	}
	for left < lengthLeft {
		(*list)[original] = leftCopy[left]
		left++
		original++
	}
	for right < lengthRight {
		(*list)[original] = rightCopy[right]
		right++
		original++
	}
}

func mergeSort(list *[]int, start, end int) {
	if start < end {
		mid := int((start + end) / 2)
		mergeSort(list, start, mid)
		mergeSort(list, mid+1, end)
		merge(list, start, mid, end)
	}
}
